File: src/handlers/users/provider/google.rs
use crate::{
    config::Config,
    handlers::users::provider::{Provider, UserProvidedData},
};
use reqwest::{blocking::Client, Error};
use serde::Deserialize;

#[derive(Deserialize)]
struct GoogleUser {
    pub email: Option<String>,
    pub name: Option<String>,
    pub verified_email: bool,
    pub picture: Option<String>,
}

#[derive(Clone)]
pub struct GoogleProvider {
    config: Config,
}

impl GoogleProvider {
    pub fn new(config: Config) -> GoogleProvider {
        return GoogleProvider { config };
    }
}

impl Provider for GoogleProvider {
    fn enabled(&self) -> bool {
        return self.config.google_enabled;
    }

    fn auth_url(&self) -> String {
        return String::from("https://accounts.google.com/o/oauth2/auth");
    }

    fn token_url(&self) -> String {
        return String::from("https://oauth2.googleapis.com/token");
    }

    fn client_id(&self) -> String {
        return self.config.google_client_id.clone().unwrap();
    }

    fn client_secret(&self) -> String {
        return self.config.google_client_secret.clone().unwrap();
    }

    fn scopes(&self) -> Vec<String> {
        return vec![String::from("email"), String::from("profile")];
    }

    fn get_user_data(&self, access_token: String) -> Result<UserProvidedData, Error> {
        let client = Client::new();

        let response = client.get("https://www.googleapis.com/oauth2/v1/userinfo?alt=json").bearer_auth(access_token).send()?;

        let data: GoogleUser = response.json()?;

        return Ok(UserProvidedData {
            email: data.email,
            verified: data.verified_email,
            avatar: data.picture,
            name: data.name,
        });
    }
}

File: handlers/provider/google.go
package provider

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/zolamk/trust/config"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
)

type googleUser struct {
	Name          *string `json:"name"`
	Email         *string `json:"email"`
	VerifiedEmail bool    `json:"verified_email"`
	Picture       *string `json:"picture"`
}

type GoogleProvider struct {
	c *config.Config
}

func NewGoogleProvider(c *config.Config) *GoogleProvider {
	return &GoogleProvider{
		c,
	}
}

func (g *GoogleProvider) name() string {
	return "google"
}

func (g *GoogleProvider) enabled() bool {
	return g.c.GoogleEnabled
}

func (g *GoogleProvider) get_config() *oauth2.Config {
	return &oauth2.Config{
		ClientID:     g.c.GoogleClientID,
		ClientSecret: g.c.GoogleClientSecret,
		Endpoint:     google.Endpoint,
		Scopes:       []string{"email", "profile"},
		RedirectURL:  fmt.Sprintf("%s/authorize/callback", g.c.InstanceURL),
	}
}

func (g *GoogleProvider) get_user_data(access_token string) (*UserData, error) {

	req, err := http.NewRequest("GET", "https://www.googleapis.com/oauth2/v1/userinfo?alt=json", nil)

	if err != nil {
		return nil, err
	}

	req.Header.Add("authorization", fmt.Sprintf("Bearer %s", access_token))

	res, err := http.DefaultClient.Do(req)

	if err != nil {
		return nil, err
	}

	decoder := json.NewDecoder(res.Body)

	google_user := &googleUser{}

	if err := decoder.Decode(google_user); err != nil {
		return nil, err
	}

	return &UserData{
		Name:     google_user.Name,
		Email:    google_user.Email,
		Avatar:   google_user.Picture,
		Verified: google_user.VerifiedEmail,
	}, nil
}
